### PR TITLE
improve readability of crate and add encode_to_slice

### PR DESCRIPTION
**Pull request by Luro02**
- moved `FromHexError` to separate file for easier reading
- moved traits into a new file (should they be separated?)
- added an `encode_to_slice` function (this can be improved with const generics in the future)
- improved the documentation a little bit

### CODE REVIEW

**Comment by LukasKalbertodt** [comment]
👍 from me for the first three commits. (Although I do like "empty" `///` lines after `/// # Example` lines.)

Regarding last commit: I don't have the time to correctly review this. Sounds useful. But maybe this commit should be in a separate PR so that this PR is only about cleanup and such things?

Also: I'm not the owner of this crate, so I don't have anything to say about merging :)

**Comment by KokaKiwi** [comment]
> Regarding last commit: I don't have the time to correctly review this. Sounds useful. But maybe this commit should be in a separate PR so that this PR is only about cleanup and such things?
Same for me, adding a new functions and moving stuff around should be separated in different PRs

File: src/lib.rs [hunk 1]
```diff
@@ -123,11 +123,13 @@ pub fn encode<T: AsRef<[u8]>>(data: T) -> String {
 // (4, 5)
 // (6, 7)
 // ...
+#[inline]
```
Comment (LukasKalbertodt) [inline src/lib.rs]: I'm pretty sure this inline attribute (and the one below) are useless. The optimizer pretty much ignores the attribute and since these functions are private and thus only used in the same crate they are defined in, adding `#[inline]` doesn't make sense. 

Out of curiosity: why did you add it?

File: src/lib.rs [hunk 2]
```diff
@@ -194,7 +196,10 @@ pub fn encode_upper<T: AsRef<[u8]>>(data: T) -> String {
 /// assert!(hex::decode("foo").is_err());
 /// ```
 pub fn decode<T: AsRef<[u8]>>(data: T) -> Result<Vec<u8>, FromHexError> {
-    FromHex::from_hex(data)
+    let mut result = vec![0; data.as_ref().len() / 2];
+    decode_to_slice(data, &mut result)?;
+
+    Ok(result)
```
Comment (LukasKalbertodt) [inline src/lib.rs]: Why replace one line of code with three more complex ones? Are there performance advantages? Otherwise I'd prefer the single line.

File: src/lib.rs (hunk 1 above)
Comment (Luro02) [reply src/lib.rs]: Well, criterion told me, that there was a performance improvement, but I am not quite sure how reliable it is. (the next run it told me, that the performance decreased by 6%)

**Comment by Luro02** [comment]
Closed in favor of #31 and #32
